internal/controllers/provisioningrequest_hwprovision.go

```diff
@@ -227,9 +227,13 @@ func (t *provisioningRequestReconcilerTask) checkNodePoolConfigStatus(ctx contex
 func (t *provisioningRequestReconcilerTask) applyNodeConfiguration(ctx context.Context, hwNodes map[string][]utils.NodeInfo,
 	nodePool *hwv1alpha1.NodePool, clusterInstance *siteconfig.ClusterInstance) error {
 
+	roleToNodeGroupName := make(map[string]string)
+	for _, nodeGroup := range nodePool.Spec.NodeGroup {
+		roleToNodeGroupName[nodeGroup.Role] = nodeGroup.Name
+	}
```

Comment:
I suggest moving this block of code to a utility function for better maintainability, and also check if the role already exists in the map.
if _, exists := roleToNodeGroupName[nodeGroup.Role]; !exists {
    	roleToNodeGroupName[nodeGroup.Role] = nodeGroup.Name
}